### PR TITLE
Primary setup for Multi-tenancy

### PR DESCRIPTION
**Pull request by dhrubo-os**
### Description

In this PR, I'm just adding the primary setup required for [Multi-tenancy feature](https://github.com/opensearch-project/ml-commons/tree/feature/multi_tenancy). After this PR, I'll start raising other PRs for adding multi-tenancy in connector, model, agents etc.

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]

### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [ ] Commits are signed per the DCO using `--signoff`.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/ml-commons/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: plugin/src/main/java/org/opensearch/ml/action/connector/GetConnectorTransportAction.java [hunk 1]
```diff
@@ -77,6 +80,15 @@ protected void doExecute(Task task, ActionRequest request, ActionListener<MLConn
                         ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.nextToken(), parser);
                         Connector mlConnector = Connector.createConnector(parser);
                         mlConnector.removeCredential();
+                        if (!Objects.equals(tenantId, mlConnector.getTenantId())) {
+                            actionListener
+                                .onFailure(
+                                    new OpenSearchStatusException(
+                                        "You don't have permission to access this connector",
```
Comment (mingshl) [inline plugin/src/main/java/org/opensearch/ml/action/connector/GetConnectorTransportAction.java:87]: When permission denied, maybe also adding the tenant id and the connector id and name to exception message, so it helps the users to get proper permission.

**Comment by mingshl** [comment]
@dhrubo-os do we have issues to explain this change?

**Comment by dhrubo-os** [comment]
> @dhrubo-os do we have issues to explain this change?

RFC: https://github.com/opensearch-project/ml-commons/issues/2358

another issue: https://github.com/opensearch-project/ml-commons/issues/2650

File: plugin/src/main/java/org/opensearch/ml/action/connector/GetConnectorTransportAction.java (hunk 1 above)
Comment (dhrubo-os) [reply plugin/src/main/java/org/opensearch/ml/action/connector/GetConnectorTransportAction.java:87]: This is a Get connector method where customer provided the connector id. So we shouldn't expose any unnecessary information from where customer can take advantage.

File: plugin/build.gradle [hunk 2]
```diff
@@ -55,6 +55,10 @@ dependencies {
 
     implementation group: 'org.opensearch', name: 'opensearch', version: "${opensearch_version}"
     implementation "org.opensearch.client:opensearch-rest-client:${opensearch_version}"
+    // Multi-tenant SDK Client
+    implementation "org.opensearch:opensearch-remote-metadata-sdk:${opensearch_version}"
+//    implementation "jakarta.json:jakarta.json-api:2.1.3"
```
[the remark this answers is not in the document]
Comment (dhrubo-os) [reply plugin/build.gradle]: done

File: plugin/build.gradle [hunk 3]
```diff
@@ -354,7 +359,16 @@ configurations.all {
     resolutionStrategy.force 'org.slf4j:slf4j-api:1.7.36'
     resolutionStrategy.force 'org.codehaus.plexus:plexus-utils:3.3.0'
     resolutionStrategy.force 'org.eclipse.platform:org.eclipse.core.runtime:3.29.0'
-    exclude group: "org.jetbrains", module: "annotations"
+
```
[the remark this answers is not in the document]
Comment (dhrubo-os) [reply plugin/build.gradle]: done

File: common/src/main/java/org/opensearch/ml/common/settings/SettingsChangeListener.java [hunk 4]
```diff
@@ -0,0 +1,10 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.ml.common.settings;
+
+public interface SettingsChangeListener {
```
[the remark this answers is not in the document]
Comment (dhrubo-os) [reply common/src/main/java/org/opensearch/ml/common/settings/SettingsChangeListener.java:11]: done

File: plugin/src/main/java/org/opensearch/ml/plugin/MachineLearningPlugin.java [hunk 5]
```diff
@@ -480,6 +494,25 @@ public Collection<Object> createComponents(
         mlIndicesHandler = new MLIndicesHandler(clusterService, client);
         encryptor = new EncryptorImpl(clusterService, client, mlIndicesHandler);
 
+        SdkClient sdkClient = SdkClientFactory
+            .createSdkClient(
+                client,
+                xContentRegistry,
+                // Here we assume remote metadata client is only used with tenant awareness.
+                // This may change in the future allowing more options for this map
+                ML_COMMONS_MULTI_TENANCY_ENABLED.get(settings)
```
Comment (brianf-aws) [inline plugin/src/main/java/org/opensearch/ml/plugin/MachineLearningPlugin.java:503]: This part is a little bit hard to follow, maybe we can extract it? I say this because you state you are going to add more in the future too.

File: plugin/src/main/java/org/opensearch/ml/utils/TenantAwareHelper.java [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    public static boolean validateTenantResource(
+        MLFeatureEnabledSetting mlFeatureEnabledSetting,
+        String tenantIdFromRequest,
+        String tenantIdFromResource,
+        ActionListener<?> listener
+    ) {
+        if (mlFeatureEnabledSetting.isMultiTenancyEnabled() && !Objects.equals(tenantIdFromRequest, tenantIdFromResource)) {
+            listener.onFailure(new OpenSearchStatusException("You don't have permission to access this resource", RestStatus.FORBIDDEN));
+            return false;
+        } else
```
Comment (brianf-aws) [inline plugin/src/main/java/org/opensearch/ml/utils/TenantAwareHelper.java:61]: same here

File: plugin/src/main/java/org/opensearch/ml/utils/TenantAwareHelper.java [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+            QueryBuilder queryBuilder = searchSourceBuilder.query();
+            if (queryBuilder instanceof TermQueryBuilder) {
+                TermQueryBuilder termQuery = (TermQueryBuilder) queryBuilder;
+                return TENANT_ID_FIELD.equals(termQuery.fieldName()); // Tenant filtering is enabled
+            }
+        }
+        return false; // Tenant filtering is not enabled
+    }
+
+    public static String getTenantID(Boolean isMultiTenancyEnabled, RestRequest restRequest) {
```
Comment (brianf-aws) [inline plugin/src/main/java/org/opensearch/ml/utils/TenantAwareHelper.java:77]: There happens to be a lot of nesting maybe lets refactor to fail early ([gaurd clause](https://www.codementor.io/@clintwinter/use-guard-clauses-for-cleaner-code-1rrsczgwxp))? 

```java
 public static String getTenantID(Boolean isMultiTenancyEnabled, RestRequest restRequest) {
   if (!isMultiTenancyEnabled)
        return null;
    Map<String, List<String>> headers = restRequest.getHeaders();
  if (headers == null)
     throw new OpenSearchStatusException("Rest request header can't be null", RestStatus.FORBIDDEN);
  if (!headers.containsKey(Constants.TENANT_ID_HEADER))
      throw new OpenSearchStatusException("Tenant ID header is missing", RestStatus.FORBIDDEN);
  List<String> tenantIdList = headers.get(Constants.TENANT_ID_HEADER);
  
   if (tenantIdList == null || tenantIdList.isEmpty()) 
         throw new OpenSearchStatusException("Tenant ID header is missing", RestStatus.FORBIDDEN);
         
  String tenantId = tenantIdList.get(0);
  if (tenantId == null)
       throw new OpenSearchStatusException("Tenant ID can't be null", RestStatus.FORBIDDEN);
               
    return tenantId
}

```

My indentation is off but the point is to check your edge cases first and let it trickle down to the happy case

**Review by brianf-aws — CHANGES_REQUESTED** [review]
I didnt get to reviewing the tests but left some comments hope they help :)

File: plugin/src/main/java/org/opensearch/ml/plugin/MachineLearningPlugin.java (hunk 5 above)
Comment (dhrubo-os) [reply plugin/src/main/java/org/opensearch/ml/plugin/MachineLearningPlugin.java:503]: We are just building the SDKClient here. Similar like [this](https://github.com/opensearch-project/flow-framework/pull/980/files#diff-9d60d6086ce87e2240d7d6bcf8d2a8b79c73400fc6ec6865f74fc571e03bf1af)

File: plugin/src/main/java/org/opensearch/ml/utils/TenantAwareHelper.java [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    /**
+     * Validates the tenant ID based on the multi-tenancy feature setting.
+     *
+     * @param mlFeatureEnabledSetting The settings that indicate whether the multi-tenancy feature is enabled.
+     * @param tenantId The tenant ID to validate.
+     * @param listener The action listener to handle failure cases.
+     * @return true if the tenant ID is valid or if multi-tenancy is not enabled; false if the tenant ID is invalid and multi-tenancy is enabled.
+     */
+    public static boolean validateTenantId(MLFeatureEnabledSetting mlFeatureEnabledSetting, String tenantId, ActionListener<?> listener) {
+        if (mlFeatureEnabledSetting.isMultiTenancyEnabled() && tenantId == null) {
```
[the remark this answers is not in the document]
Comment (dhrubo-os) [reply plugin/src/main/java/org/opensearch/ml/utils/TenantAwareHelper.java:36]: From the plugin level, we don't have visibility if the tenant id is valid or not. We perform this validation from the interceptor level. Here we just check if multi-tenancy is enabled then tenant id needs to be present.

File: plugin/src/main/java/org/opensearch/ml/utils/TenantAwareHelper.java [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    public static String getTenantID(Boolean isMultiTenancyEnabled, RestRequest restRequest) {
+        if (isMultiTenancyEnabled) {
+            Map<String, List<String>> headers = restRequest.getHeaders();
+            if (headers != null) {
+                if (headers.containsKey(Constants.TENANT_ID_HEADER)) {
+                    List<String> tenantIdList = headers.get(Constants.TENANT_ID_HEADER);
+                    if (tenantIdList != null && !tenantIdList.isEmpty()) {
+                        String tenantId = tenantIdList.get(0);
```
[the remark this answers is not in the document]
Comment (dhrubo-os) [reply plugin/src/main/java/org/opensearch/ml/utils/TenantAwareHelper.java:84]: HTTP headers are represented as a Map<String, List<String>> because the HTTP protocol allows multiple values for a single header key (e.g., Set-Cookie). Even though most headers, like Tenant-ID, typically have only one value, the list structure ensures flexibility and compatibility. In practice, we usually retrieve the first value (list.get(0)) since only one value is needed in most cases.

File: plugin/src/main/java/org/opensearch/ml/settings/MLCommonsSettings.java [hunk 10]
```diff
@@ -263,4 +268,35 @@ private MLCommonsSettings() {}
...
+    // OPENSEARCH_JAVA_OPTS="-Dplugins.ml_commons.multi_tenancy_enabled=true" ./bin/opensearch
+    // 3. In the opensearch.yml configuration file:
+    // plugins.ml_commons.multi_tenancy_enabled: true
+    // After setting it, a full cluster restart is required for the changes to take effect.
+    public static final Setting<Boolean> ML_COMMONS_MULTI_TENANCY_ENABLED = Setting
+        .boolSetting("plugins.ml_commons.multi_tenancy_enabled", false, Setting.Property.NodeScope);
+
+    /** This setting sets the remote metadata type */
+    public static final Setting<String> REMOTE_METADATA_TYPE = Setting
+        .simpleString("plugins.ml_commons." + REMOTE_METADATA_TYPE_KEY, Setting.Property.NodeScope, Setting.Property.Final);
```
[the remark this answers is not in the document]
Comment (dhrubo-os) [reply plugin/src/main/java/org/opensearch/ml/settings/MLCommonsSettings.java:289]: Nice!!

File: plugin/src/main/java/org/opensearch/ml/utils/TenantAwareHelper.java (hunk 7 above)
Comment (dhrubo-os) [reply plugin/src/main/java/org/opensearch/ml/utils/TenantAwareHelper.java:77]: Yeah this is better.

**Comment by dhrubo-os** [comment]
```
Tests with failures:
 - org.opensearch.ml.tools.VisualizationsToolIT.testVisualizationNotFound
 - org.opensearch.ml.tools.VisualizationsToolIT.testVisualizationFound
111 tests completed, 2 failed, 10 skipped
```


Same flaky errors

File: ml-algorithms/build.gradle [hunk 11]
```diff
@@ -87,8 +87,10 @@ lombok {
 configurations.all {
     resolutionStrategy.force 'com.google.protobuf:protobuf-java:3.25.5'
     resolutionStrategy.force 'org.apache.commons:commons-compress:1.26.0'
+    resolutionStrategy.force 'software.amazon.awssdk:bom:2.29.12'
```
Comment (b4sjoo) [inline ml-algorithms/build.gradle:90]: Do you know why use a force here? If the version matter then you should be aware that it will cause version conflict when goes to the managed service.

Comment (dhrubo-os) [reply ml-algorithms/build.gradle:90]: refer to this: https://github.com/opensearch-project/opensearch-remote-metadata-sdk/pull/25#issuecomment-2580952651

**Review by mingshl — APPROVED** [review]

File: plugin/src/main/java/org/opensearch/ml/utils/TenantAwareHelper.java (hunk 7 above)
Comment (dbwiddis) [reply plugin/src/main/java/org/opensearch/ml/utils/TenantAwareHelper.java:77]: > if (headers == null)
>     throw new OpenSearchStatusException("Rest request header can't be null", RestStatus.FORBIDDEN);

FYI, this case is impossible.  I tried to write a unit test for the Flow Framework version of this util and could not (short of using reflection) make the headers null.  The only way to make it null is using a constructor, where it passes through a `headers.get()` to get the content type from the header; and if that didn't stop it, `Collections.unmodifiableMap()` [would throw an NPE on null headers](https://github.com/AdoptOpenJDK/openjdk-jdk11/blob/19fb8f93c59dfd791f62d41f332db9e306bc1422/src/java.base/share/classes/java/util/Collections.java#L1443-L1444) as well.

**Review by b4sjoo — APPROVED** [review]

File: plugin/src/main/java/org/opensearch/ml/utils/TenantAwareHelper.java (hunk 7 above)
Comment (dhrubo-os) [reply plugin/src/main/java/org/opensearch/ml/utils/TenantAwareHelper.java:77]: Yeah, that's a good point! I agree but I prefer to still keep this as a defensive coding practice.
